### PR TITLE
Support Open-Generic-Types in 'nameof' expressions.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/4827

I've run this by @MadsTorgersen and he also think this is something that we should probably support.

Note: this PR is out in order for me to find out if i've missed anything.  This will not be merged in until it has both PR signoff and has signoff from the language design team.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
reviewers: @AlekseyTs @VSadov @agocke @jaredpar @KevinRansom

File: src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs [hunk 1]
```diff
@@ -8,78 +8,25 @@
...
-    /// should be considered open so that the flag can be set 
-    /// appropriately in BoundTypeOfOperator.
+    /// Base type for nameof and typeof expression arguments.  Properly handles binding open types
+    /// in the argument.
     /// </summary>
-    internal sealed class TypeofBinder : Binder
+    internal abstract class TypeofOrNameofBinder : Binder
     {
         private readonly Dictionary<GenericNameSyntax, bool> _allowedMap;
-        private readonly bool _isTypeExpressionOpen;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/TypeofBinder.cs]: This field was never used.  So i removed it entirely.

**Comment by AlekseyTs** [comment]
@gafter FYI

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 2]
```diff
@@ -847,6 +847,15 @@ private NamedTypeSymbol ConstructNamedTypeUnlessTypeArgumentOmitted(CSharpSyntax
         {
             if (typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument))
             {
+                if (typeSyntax.Kind() == SyntaxKind.GenericName)
+                {
+                    var genericName = (GenericNameSyntax)typeSyntax;
+                    if (genericName.IsUnboundGenericName && IsUnboundTypeAllowed(genericName))
+                    {
+                        return type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: This doesn't feel right, we shouldn't return the same type symbol.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 3]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+        [Fact, WorkItem(4827, "https://github.com/dotnet/roslyn/issues/4827")]
+        public void NameofOpenGenericType1()
+        {
+            var source =
+@"class A<T> { }
+class Test {
+  void M(string s = nameof(A<>)) {
+  }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Consider verifying behavior

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 4]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+        [Fact, WorkItem(4827, "https://github.com/dotnet/roslyn/issues/4827")]
+        public void NameofOpenGenericType4()
+        {
+            var source =
+@"class B<X,Y> { }
+class Test {
+  void M(string v = nameof(B<,>)) {
+  }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Consider verifying behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: How is the best way to do that?

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Why not?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 5]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+                Diagnostic(ErrorCode.ERR_BadArity, "C").WithArguments("C<T>", "type", "1").WithLocation(3, 28));
+        }
+
+        [Fact, WorkItem(4827, "https://github.com/dotnet/roslyn/issues/4827")]
+        public void NameofOpenGenericType6()
+        {
+            var source =
+@"class C<T> { public class Nested { } }
+class Test {
+  void M(string s = nameof(C<>.Nested)) {
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: What does SemanticModel returns for the argument of nameof and its subexpressions

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 6]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+  }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Consider verifying behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 7]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+
+        [Fact, WorkItem(4827, "https://github.com/dotnet/roslyn/issues/4827")]
+        public void NameofOpenGenericType_OpenNameOnRight()
+        {
+            var source =
+@"class Test {
+  void M(string s = nameof(System.Action<>)) {
+  }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Consider verifying behavior.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Will add tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs [hunk 8]
```diff
@@ -1168,5 +1162,156 @@ public void NameofTypeParameterInParameterInitializer()
...
+        public void NameofOpenGenericTypeWithMethod()
+        {
+            var source =
+@"
+using System.Collections.Generic;
+class Test {
+  void M(string s = nameof(List<>.Add)) {
+  }
+}";
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(source).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Consider verifying behavior.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Because it is an unbound type

**Comment by AlekseyTs** [comment]
Please provide details about proposed spec change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs]: Have the program print the value for example and check the output.

**Comment by AlekseyTs** [comment]
Should something change for VB?

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs @MadsTorgersen and i will have to work through the spec change.  According to him, we're somewhat underspecified now.  So we'll use this as an opportunity to fully spec out what is allowed in a ```nameof``` argument.  For purposes of this PR, your intuition should be "we now allow the same 'open generic types' that we would allow in a ```typeof``` argument".

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs 

> Should something change for VB?

I think it would be off if this was approved for C# only.

**Comment by gafter** [comment]
We need a proposed language specification before I can even review whether or not this implements the proposal. I wonder what the required API behavior is for "half-constructed" types...

```cs
class G<T,U>
{
    public KeyValuePair<T,U> AProperty { get; }
}
class D<T> : G<T, int>
{
}
... nameof(D<>.AProperty) ...
```

In this case what should the compiler's API report about the type of the symbol in the nameof node?

**Comment by CyrusNajmabadi** [comment]
@gafter The purpose of this PR is to drive toward the proposed language specific change.

**Comment by AlekseyTs** [comment]
Perhaps a change for VB should be included as well.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs 

> Perhaps a change for VB should be included as well.

I'd rather not go through that work if this proposal is not approved by the LDM.  This is intended to be a strawman.

**Comment by jaredpar** [comment]
@CyrusNajmabadi if the implementation is supposed to inform the decision then VB support needs to be present as well.  Otherwise I don't see the value of having a PR here.  Either the implementation helps with the decision and we need both languages or it doesn't and there is no need for a PR.

**Comment by CyrusNajmabadi** [comment]
@gafter 

>  I wonder what the required API behavior is for "half-constructed" types...

Test added to demonstrate the behavior i think is currently desired.  I'll include this as an important case in the MD i write up on the topic.

**Comment by CyrusNajmabadi** [comment]
Closing so that this can be processed like a normal "feature" PR.

**Pull request closed without merging**